### PR TITLE
Add optimizer to reservation units

### PR DESCRIPTION
**Pull request by matti-lamppu**
## 🛠️ Changelog
[//]: # "Describe the changes in this pull request here."

- Refactor all endpoints to new object types. Adds SQL optimizer to all endpoints.
- Refactor all mutations to new mutation classes. Also change most serializers to NestingModelSerializer, except reservation stuff, which requires another refactor later
- Convert to new permissions classes. These should have the same functionality, but are more streamlined, especially for mutations (where previously it was not possible to have separate permission checks for create/update/delete)
- Add `mutations.py` and `queries.py` files where all object types are imported first before importing them to `schema.py`. This is to avoid issues with creating mutation classes (see explanation in `mutations.py` and `queries.py`)
- Add a new filter to reservation unit endpoint: `calculateFirstReservableTime`. This should be set to `true` to enable the first reservable time calculation. This is to make the endpoint more performant if the information is not required. Querying the first reservable time fields without setting this to `true` will result in an error.
- Change how GraphQL file uploads work to be more compliant with [graphql-multipart-request-spec](https://github.com/jaydenseric/graphql-multipart-request-spec).
- Refactors all remaining graphql tests to functional tests under the tests folder... so. many. tests...

### Other changes

- Allow staff reservations to be created for the current day (or the previous day if it's the first hour of the day)
- Add two new filters to the units endpoint: `only_direct_bookable` & `only_seasonal_bookable`, which filter units to those that contain at least one reservation unit that is direct or seasonal bookable respectively.

## 🧪 Test plan
[//]: # "Help your fellow reviewer and write a short description of what's the fastest way to test your changes."

- Automated tests

## 🎫 Tickets
[//]: # "This pull request resolves all or part of the following ticket(s)"

- [TILA-2959](https://helsinkisolutionoffice.atlassian.net/browse/TILA-2959), [TILA-2960](https://helsinkisolutionoffice.atlassian.net/browse/TILA-2960), [TILA-2961](https://helsinkisolutionoffice.atlassian.net/browse/TILA-2961), [TILA-3246](https://helsinkisolutionoffice.atlassian.net/browse/TILA-3246), [TILA-3251](https://helsinkisolutionoffice.atlassian.net/browse/TILA-3251)

[TILA-2959]: https://helsinkisolutionoffice.atlassian.net/browse/TILA-2959?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by matti-lamppu** [comment]
@vergama Just check the essentials, like that permissions and translations are correct. Most stuff should not have changed. I've put things in [separate commits](https://github.com/City-of-Helsinki/tilavarauspalvelu-core/pull/1131/commits) to make this easier.

@ranta Let's go over this in more detail on a call.

File: locale/fi/LC_MESSAGES/django.po [hunk 1]
```diff
@@ -2327,190 +2316,337 @@ msgstr "Ylin hinta"
...
+msgid "Applying free of charge"
+msgstr "Hakee alennusta"
+
+#: reservations/admin.py
+msgid "Free of charge reason"
+msgstr "Alennuksen haun syy"
+
+#: reservations/admin.py
+msgid "Reservee ID"
+msgstr "Varaajan tunniste"
```
Comment (vergama) [inline locale/fi/LC_MESSAGES/django.po:2377]: Varaajan y-tunnus

File: locale/fi/LC_MESSAGES/django.po [hunk 2]
```diff
@@ -2327,190 +2316,337 @@ msgstr "Ylin hinta"
...
+msgid "Additional details for denying or approving the reservation"
+msgstr "Lisätietoja varauksen hylkäämiseen tai hyväksymiseen"
+
+#: reservations/admin.py
+msgid "Working memo for staff users"
+msgstr "Muistio henkilökunnalle"
+
+#: reservations/admin.py
+msgid "Reservation begin date and time"
+msgstr "Varauksen alkamispäivä"
```
Comment (vergama) [inline locale/fi/LC_MESSAGES/django.po:2501]: Varauksen alkamishetki

File: locale/fi/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -2327,190 +2316,337 @@ msgstr "Ylin hinta"
...
+
+#: reservations/admin.py
+msgid "Reservation end date and time"
+msgstr "Varauksen loppumispäivä"
```
Comment (vergama) [inline locale/fi/LC_MESSAGES/django.po:2505]: Varauksen loppumishetki

File: locale/fi/LC_MESSAGES/django.po [hunk 4]
```diff
@@ -2327,190 +2316,337 @@ msgstr "Ylin hinta"
...
 msgid "Reason for applying for a free-of-charge reservation"
 msgstr "Syy alennuskelpoisuuden hakemiselle"
 
-#: reservations/models/reservation.py
-#: reservations/models/reservation_statistic.py
-msgid "State"
-msgstr "Tila"
+#: reservations/admin.py
+msgid "Reservee's business or association identity code"
+msgstr "Varaaja yrityksen tai yhdistyksen tunnus"
```
Comment (vergama) [inline locale/fi/LC_MESSAGES/django.po:2553]: Varaajan ... y-tunnus

File: locale/fi/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -2327,190 +2316,337 @@ msgstr "Ylin hinta"
...
-#: reservations/models/reservation.py
-msgid "Handling details for this reservation"
-msgstr "Käsittelymuistio tälle varaukselle"
+#: reservations/admin.py
+msgid "Purpose of the reservation"
+msgstr "Varauksen käyttötarkoitus"
+
+#: reservations/admin.py
+msgid "Home city of the group or association"
+msgstr "Yhdistyksen tai ryhmän kotikunta"
```
Comment (vergama) [inline locale/fi/LC_MESSAGES/django.po:2609]: In the case of reservations:
"Reservee's home city" = "Varaajan kotikunta"

**Review by vergama — APPROVED** [review]
'S’all good, man'

**Comment by matti-lamppu** [comment]
Flaky test: https://github.com/City-of-Helsinki/tilavarauspalvelu-core/actions/runs/8434831276/job/23098918644#step:15:392

**Comment by matti-lamppu** [comment]
Merged PRs #1151, #1150, #1149, #1146, #1145, and #1143 to this PR, since they should be merged together.

**Comment by matti-lamppu** [comment]
@joonatank Gave the go-ahead to merge this. Frontend to follow shortly.

**Pull request merged**